### PR TITLE
Add image classification script, no trainer

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds an example script for image classification that leverages Accelerate instead of the HuggingFace Trainer.

To do:

- [x] verify local `train_dir` and `validation_dir`
- [x] update README
- [x] add log fixes (Tensorboard)

Both can be updated after #16585 is merged.

### CODE REVIEW

**Comment by NielsRogge** [comment]
@sgugger not sure why, but the test for the script fails:
```

WARNING  datasets.builder:builder.py:388 Using custom data configuration huggingface--image-classification-test-sample-b7448dc7ae37f2cf
INFO     run_image_classification_no_trainer:run_image_classification_no_trainer.py:388 ***** Running training *****
INFO     run_image_classification_no_trainer:run_image_classification_no_trainer.py:389   Num examples = 8
INFO     run_image_classification_no_trainer:run_image_classification_no_trainer.py:390   Num Epochs = 3
INFO     run_image_classification_no_trainer:run_image_classification_no_trainer.py:391   Instantaneous batch size per device = 2
INFO     run_image_classification_no_trainer:run_image_classification_no_trainer.py:392   Total train batch size (w. parallel, distributed & accumulation) = 2
INFO     run_image_classification_no_trainer:run_image_classification_no_trainer.py:393   Gradient Accumulation steps = 1
INFO     run_image_classification_no_trainer:run_image_classification_no_trainer.py:394   Total optimization steps = 12
INFO     run_image_classification_no_trainer:run_image_classification_no_trainer.py:471 epoch 0: {'accuracy': 0.0}
INFO     run_image_classification_no_trainer:run_image_classification_no_trainer.py:471 epoch 1: {'accuracy': 0.0}
INFO     run_image_classification_no_trainer:run_image_classification_no_trainer.py:471 epoch 2: {'accuracy': 0.0}
```
Weirdly, it passes locally for me.

File: examples/pytorch/image-classification/run_image_classification_no_trainer.py [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+        dataset["train"] = split["train"]
+        dataset["validation"] = split["test"]
+
+    # Prepare label mappings.
+    # We'll include these in the model's config to get human readable labels in the Inference API.
+    labels = dataset["train"].features["labels"].names
+    label2id, id2label = dict(), dict()
+    for i, label in enumerate(labels):
+        label2id[label] = str(i)
+        id2label[str(i)] = label
```
Comment (sgugger) [inline examples/pytorch/image-classification/run_image_classification_no_trainer.py:261]:
```suggestion
    label2id = {label: str(i) for i, label in enumerate(labels)}
    id2label = {str(i): label for i, label in enumerate(labels)}
```
Two list comprehensions will always be faster than one manual loop :-)

File: examples/pytorch/image-classification/run_image_classification_no_trainer.py [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+
+    # Load pretrained model and feature extractor
+    #
+    # In distributed training, the .from_pretrained methods guarantee that only one local process can concurrently
+    # download model & vocab.
+    config = AutoConfig.from_pretrained(
+        args.model_name_or_path,
+        num_labels=len(id2label),
```
Comment (sgugger) [inline examples/pytorch/image-classification/run_image_classification_no_trainer.py:269]:
```suggestion
```
Feel free to ignore, but passing `num_labels` is not necessary if you pass `id2label` and `label2id`.

File: examples/pytorch/test_accelerate_examples.py [hunk 3]
```diff
@@ -321,3 +322,24 @@ def test_run_semantic_segmentation_no_trainer(self):
             run_semantic_segmentation_no_trainer.main()
             result = get_results(tmp_dir)
             self.assertGreaterEqual(result["eval_overall_accuracy"], 0.10)
+
+    def test_image_classification_no_trainer(self):
```
Comment (sgugger) [inline examples/pytorch/test_accelerate_examples.py:326]:
```suggestion
    def test_run_image_classification_no_trainer(self):
```
For consistency with other tests.

File: examples/pytorch/test_accelerate_examples.py [hunk 4]
```diff
@@ -321,3 +322,24 @@ def test_run_semantic_segmentation_no_trainer(self):
...
+        tmp_dir = self.get_auto_remove_tmp_dir()
+        testargs = f"""
+            run_image_classification_no_trainer.py
+            --dataset_name huggingface/image-classification-test-sample
+            --output_dir {tmp_dir}
+            --num_warmup_steps=8
+            --learning_rate=3e-3
+            --per_device_train_batch_size=2
+            --per_device_eval_batch_size=1
+            --checkpointing_steps epoch
```
Comment (sgugger) [inline examples/pytorch/test_accelerate_examples.py:336]:
```suggestion
            --checkpointing_steps epoch
            --seed 42
```
This helps with making the test locally pass for me.

**Review by sgugger — APPROVED** [review]
Nice new addition!
Tested locally and had the test pass one time out of 4, so maybe it's a seed issue? Setting a seed made it pass several times in a row. The learning rate seems extremely high for a Transformers model (but maybe ViT exepects that) so I'd say the test is a bit unstable without setting the seed.

**Comment by NielsRogge** [comment]
I'm getting issues when only passing `id2label` and `label2id` to the config, but not the `num_labels`:

```
if size_average is not None or reduce is not None:
            reduction = _Reduction.legacy_get_string(size_average, reduce)
>       return torch._C._nn.cross_entropy_loss(input, target, weight, _Reduction.get_enum(reduction), ignore_index, label_smoothing)
E       IndexError: Target 6 is out of bounds.
```

**Comment by sgugger** [comment]
Oh ok, shouldn't be the case. Let's put back the `num_labels` for now and I'll have a look later at why it failed to update properly.

**Pull request merged**